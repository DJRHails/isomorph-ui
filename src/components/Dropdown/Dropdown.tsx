import React, { FC } from 'react';
import { UseDropdownProps, DropdownProvider, useDropdown } from './useDropdown';
import { ReactNodeOrRenderProp } from '/types/utils';
import { runIfFn } from '/utils/function';

export interface DropdownProps extends UseDropdownProps {
  children: ReactNodeOrRenderProp<{ isOpen: boolean; onClose(): void }>
} 

/**
 * Dropdown provides context, state and focus management.
 * Doesn't render DOM node.
 */
export const Dropdown: FC<DropdownProps> = (props: DropdownProps) => {
  const context = useDropdown(props);

  return (
    <DropdownProvider value={context}>
      {runIfFn(props.children, {
        isOpen: context.isOpen,
        onClose: context.onClose
      })}
    </DropdownProvider>
  );
};
Dropdown.displayName = 'Dropdown';