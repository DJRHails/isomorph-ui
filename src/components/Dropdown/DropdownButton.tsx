import React, { forwardRef, Ref, ReactNode } from 'react';
import { mergeRefs } from '/utils';
import styled from 'styled-components';
import { css } from '@styled-system/css';
import { Box } from '../Box';
import { useDropdownButton, UseDropdownButtonProps } from './useDropdownButton';

export interface DropdownButtonProps extends UseDropdownButtonProps, Record<string, any>{
  as?: any
  children?: ReactNode
  /**
   * Additional props to pass to as
   */
}

const StyledDropdownButton = styled(Box)(
  css({
    display: 'inline-flex',
    appearance: 'none',
    alignItems: 'center',
    outline: 0,
    transition: 'all 250ms',
  })
);

/**
 * The trigger for the menu list. Must be a direct child of `Dropdown`.
 */
export const DropdownButton = forwardRef((
  props: DropdownButtonProps,
  ref: Ref<any>,
) => {
  const { children, as: Comp, ...otherProps } = props;

  const ownProps = useDropdownButton(otherProps);
  const ownRef = mergeRefs(ref, ownProps.ref);

  const Wrapper = Comp || StyledDropdownButton;

  return (
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    //@ts-ignore
    <Wrapper {...ownProps} ref={ownRef}>
      <Box
        as="span"
        sx={{
          pointerEvents: 'none',
          flex: '1',
        }}
      >
        {children}
      </Box>
    </Wrapper>
  );
});

DropdownButton.displayName = 'DropdownButton';