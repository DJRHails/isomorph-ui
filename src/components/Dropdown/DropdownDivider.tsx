import { css } from '@styled-system/css';
import styled from 'styled-components';

export interface DropdownDividerProps {}

export const DropdownDivider = styled('hr')(
  css({
    border: 0,
    borderBottom: '1px solid',
    borderColor: 'hover',
    my: '0.5rem',
    opacity: 0.6,
  })
);

DropdownDivider.displayName = 'DropdownDivider';
