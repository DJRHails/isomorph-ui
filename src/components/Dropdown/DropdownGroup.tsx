import React, { ReactNode } from 'react';
import { Box } from '../Box';
import { css } from '@styled-system/css';
import styled from 'styled-components';

export interface DropdownGroupProps {
  title?: string
  children?: ReactNode
}

const StyledTitle = styled('p')(
  css({
    mx: 4,
    my: 2,
    fontWeight: 'bold',
    fontSize: '0.875rem'
  })
);

export const DropdownGroup = (props: DropdownGroupProps) => {
  const { title, children } = props;

  return (
    <Box role="group">
      {title && (
        <StyledTitle>
          {title}
        </StyledTitle>
      )}
      {children}
    </Box>
  );
};

DropdownGroup.displayName = 'DropdownGroup';
