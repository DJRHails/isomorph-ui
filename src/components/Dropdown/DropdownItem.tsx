import styled from 'styled-components';
import React, { ReactElement, forwardRef, ReactNode } from 'react';
import { mergeRefs } from '/utils';
import { SpaceProps } from 'styled-system';
import { css } from '@styled-system/css';
import { UseDropdownItemProps, useDropdownItem } from './useDropdownItem';

const StyledDropdownItem = styled('button')(
  css({
    textDecoration: 'none',
    color: 'inherit',
    userSelect: 'none',
    display: 'flex',
    width: '100%',
    alignItems: 'center',
    textAlign: 'left',
    flex: '0 0 auto',
    outline: 0,
    px: 2,
    py: 3,
    '&:hover, &:focus': {
      bg: 'hover'
    }
  })
);

interface DropdownItemOptions extends Omit<UseDropdownItemProps, 'context'> {
  /**
   * The icon to render before the menu item's label.
   */
  icon?: ReactElement
  /**
   * The spacing between the icon and menu item's label
   */
  iconSpacing?: SpaceProps['mr']
  /**
   * Right-aligned label text content, useful for displaying hotkeys.
   */
  command?: string

  children?: ReactNode
}
export interface DropdownItemProps extends DropdownItemOptions {
  children?: ReactNode
}

export const DropdownItem = forwardRef((
  props: DropdownItemProps,
  ref,
) => {
  const {
    children,
    ...otherProps
  } = props;

  const ownProps = useDropdownItem(otherProps);
  const ownRef = mergeRefs(ownProps.ref, ref);

  return (
    <StyledDropdownItem {...ownProps} ref={ownRef}>
      {children}
    </StyledDropdownItem>
  );
});

DropdownItem.displayName = 'DropdownItem';