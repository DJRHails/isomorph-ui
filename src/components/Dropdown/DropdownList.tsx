import { Box } from '../Box';
import React, { Ref } from 'react';
import { useDropdownList, UseDropdownListProps } from './useDropdownList';
import { mergeRefs } from '/utils';

export interface DropdownListProps extends UseDropdownListProps {
  
}

export const DropdownList = React.forwardRef((
  props: DropdownListProps,
  ref: Ref<any>,
) => {
  const menulist = useDropdownList(props);
  return (
    <Box
      {...menulist as any}
      ref={mergeRefs(menulist.ref, ref)}
      sx={{
        outline: 0,
        boxShadow: 'md',
        color: 'inherit',
        minWidth: '56',
        py: 2,
        zIndex: 1,
        bg: 'background',
        borderRadius: 'md',
        borderWidth: '1px',
        borderColor: 'hover',
        ...menulist.style
      }}
    />
  );
});

DropdownList.displayName = 'DropdownList';