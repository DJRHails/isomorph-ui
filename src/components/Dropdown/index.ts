export type { DropdownProps } from './Dropdown';
export { Dropdown } from './Dropdown';

export type { DropdownListProps } from './DropdownList';
export { DropdownList } from './DropdownList';

export type { DropdownButtonProps } from './DropdownButton';
export { DropdownButton } from './DropdownButton';

export type { DropdownDividerProps } from './DropdownDivider';
export { DropdownDivider } from './DropdownDivider';

export type { DropdownGroupProps } from './DropdownGroup';
export { DropdownGroup } from './DropdownGroup';