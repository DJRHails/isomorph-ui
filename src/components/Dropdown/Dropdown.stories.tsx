import React from 'react';
import { KindMeta } from '/types/storybook';

import { Dropdown } from './Dropdown';
import { Button } from '../Button';
import { DropdownButton } from './DropdownButton';
import { DropdownList } from './DropdownList';
import { DropdownGroup } from './DropdownGroup';
import { Portal } from '..';
import { DropdownItem } from './DropdownItem';
import { DropdownDivider } from './DropdownDivider';

const story: KindMeta<typeof Dropdown> = {
  title: 'Components/Dropdown',
  component: Dropdown,
};

const words = [
  'About Visual Studio Code',
  'Check for updates',
  'Preferences',
  'Services',
  'Hide Visual Studio Code',
  'Show All',
];

export const simple = () => (
  <div style={{ minHeight: 4000 }}>
    <Dropdown>
      <DropdownButton
        // rightIcon={<FaUnlink />}
      >
        <Button>Words</Button>
      </DropdownButton>
      <DropdownList>
        Word
      </DropdownList>
    </Dropdown>
  </div>
);


export const Basic = () => (
  <div style={{ minHeight: 4000 }}>
    <Dropdown isLazy>
      <DropdownButton>
        Open Word menu
      </DropdownButton>
      <DropdownList>
        {words.map((word) => (
          <DropdownItem key={word}>{word}</DropdownItem>
        ))}
      </DropdownList>
    </Dropdown>
  </div>
);

export const WithPortal = () => (
  <Dropdown>
    <DropdownButton as={Button} variant="solid" buttonColor="green" size="sm">
      Open menu
    </DropdownButton>
    <Portal>
      <DropdownList>
        <DropdownItem>Dropdown 1</DropdownItem>
        <DropdownItem>Dropdown 2</DropdownItem>
        <DropdownItem>Dropdown 3</DropdownItem>
        <DropdownItem>Dropdown 4</DropdownItem>
      </DropdownList>
    </Portal>
  </Dropdown>
);

export const withGroupedItems = () => (
  <Dropdown>
    {({ isOpen }) => (
      <>
        <DropdownButton as={Button} variant="solid" buttonColor="green" size="sm">
          {isOpen ? 'Close' : 'Open'} menu
        </DropdownButton>
        <DropdownList>
          <DropdownGroup title="Group 1">
            <DropdownItem>Share...</DropdownItem>
            <DropdownItem>Move...</DropdownItem>
          </DropdownGroup>
          <DropdownDivider/>
          <DropdownGroup title="Group 2">
            <DropdownItem isDisabled>Rename...</DropdownItem>
            <DropdownItem>Delete...</DropdownItem>
          </DropdownGroup>
        </DropdownList>
      </>
    )}
  </Dropdown>
);

export default story;