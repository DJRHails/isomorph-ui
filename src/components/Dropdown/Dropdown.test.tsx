import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import theme from '/themes/theme';

import { Dropdown } from './Dropdown';
import { DropdownButton } from './DropdownButton';
import { DropdownList } from './DropdownList';
import { DropdownItem } from './DropdownItem';

const DefaultDropdown = () => (
  <Dropdown isLazy>
    <DropdownButton>
        Open dropdown
    </DropdownButton>
    <DropdownList>
      <DropdownItem>A word</DropdownItem>
    </DropdownList>
  </Dropdown>
);

describe('Dropdown component sanity', () => {
  it('has name', () => {
    expect(Dropdown.displayName).toBe('Dropdown');
  });

  // TODO(DJRHails): Fix as getting unknown update request
  // it('matches default snapshot', () => {
  //   const component = render(<DefaultDropdown />);
  //   expect(component).toMatchSnapshot();
  // });

  // it('matches themed snapshot', () => {
  //   const component = render(
  //     <ThemeProvider theme={theme}>
  //       <DefaultDropdown />
  //     </ThemeProvider>
  //   );
  //   expect(component).toMatchSnapshot();
  // });
});