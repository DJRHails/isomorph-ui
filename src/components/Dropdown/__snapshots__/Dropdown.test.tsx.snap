// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component sanity matches default snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        aria-controls="dropdown-list-2"
        aria-expanded="false"
        aria-haspopup="dropdown"
        class="sc-AxjAm sc-AxirZ jdWLdj"
        id="dropdown-button-2"
      >
        <span
          class="sc-AxjAm gKHFTW"
        >
          Open dropdown
        </span>
      </div>
      <div
        aria-orientation="vertical"
        class="sc-AxjAm dgHzaX"
        data-placement="bottom-start"
        hidden=""
        id="dropdown-list-2"
        role="dropdown"
        style="transform-origin: top left;"
        tabindex="-1"
      />
    </div>
  </body>,
  "container": <div>
    <div
      aria-controls="dropdown-list-2"
      aria-expanded="false"
      aria-haspopup="dropdown"
      class="sc-AxjAm sc-AxirZ jdWLdj"
      id="dropdown-button-2"
    >
      <span
        class="sc-AxjAm gKHFTW"
      >
        Open dropdown
      </span>
    </div>
    <div
      aria-orientation="vertical"
      class="sc-AxjAm dgHzaX"
      data-placement="bottom-start"
      hidden=""
      id="dropdown-list-2"
      role="dropdown"
      style="transform-origin: top left;"
      tabindex="-1"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Dropdown component sanity matches themed snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        aria-controls="dropdown-list-12"
        aria-expanded="false"
        aria-haspopup="dropdown"
        class="sc-AxjAm sc-AxirZ jdWLdj"
        id="dropdown-button-12"
      >
        <span
          class="sc-AxjAm gKHFTW"
        >
          Open dropdown
        </span>
      </div>
      <div
        aria-orientation="vertical"
        class="sc-AxjAm jBJqQZ"
        data-placement="bottom-start"
        hidden=""
        id="dropdown-list-12"
        role="dropdown"
        style="transform-origin: top left;"
        tabindex="-1"
      />
    </div>
  </body>,
  "container": <div>
    <div
      aria-controls="dropdown-list-12"
      aria-expanded="false"
      aria-haspopup="dropdown"
      class="sc-AxjAm sc-AxirZ jdWLdj"
      id="dropdown-button-12"
    >
      <span
        class="sc-AxjAm gKHFTW"
      >
        Open dropdown
      </span>
    </div>
    <div
      aria-orientation="vertical"
      class="sc-AxjAm jBJqQZ"
      data-placement="bottom-start"
      hidden=""
      id="dropdown-list-12"
      role="dropdown"
      style="transform-origin: top left;"
      tabindex="-1"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
